src/sidebar/store/use-store.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+/* global process */
+
+const shallowEqual = require('shallowequal');
+const { useEffect, useRef, useReducer } = require('preact/hooks');
+
+const { useService } = require('../util/service-context');
+const warnOnce = require('../util/warn-once');
+
+/**
+ * Hook for accessing state or actions from the store inside a component.
+ *
+ * This hook fetches the store using `useService` and returns the result of
+ * passing it to the provided callback. The callback will be re-run whenever
+ * the store updates and the component will be re-rendered if the result of
+ * `callback(store)` changed.
+ *
+ * This ensures that the component updates when relevant store state changes.
+ *
+ * @example
+ *   function MyWidget({ widgetId }) {
+ *     const widget = useStore(
+ *      useCallback(store => store.getWidget(widgetId), [widgetId])
+ *     );
+ *     const hideWidget = useStore(store => store.hideWidget);
+ *
+ *     return (
+ *       <div>
+ *         {widget.name}
+ *         <button onClick={() => hideWidget(widgetId)}>Hide</button>
+ *       </div>
+ *     )
+ *   }
+ *
+ * @template T
+ * @param {Function} callback -
+ *   Callback that receives the store as an argument and returns some state
+ *   and/or actions extracted from the store.
+ * @return {T} - The result of `callback(store)`
+ */
+function useStore(callback) {
+  const store = useService('store');
+
+  // Store the last-used callback in a ref so we can access it in the effect
+  // below without having to re-subscribe to the store when it changes.
+  const lastCallback = useRef(null);
+  lastCallback.current = callback;
+
+  const lastResult = useRef(null);
+  lastResult.current = callback(store);
+
+  // Check for a performance issue caused by `callback` returning a different
+  // result on every call, even if the store has not changed.
+  if (process.env.NODE_ENV !== 'production') {
+    if (!shallowEqual(lastResult.current, callback(store))) {
+      warnOnce(
+        'The output of a callback passed to `useStore` changes every time. ' +
+          'This will lead to a component updating more often than necessary.'
+      );
+    }
+  }
+
+  // Abuse `useReducer` to force updates when the store state changes.
+  const [, forceUpdate] = useReducer(x => x + 1, 0);
+
+  // Connect to the store, call `callback(store)` whenever the store changes
+  // and re-render the component if the result changed.
+  useEffect(() => {
+    function checkForUpdate() {
+      const result = lastCallback.current(store);
+      if (shallowEqual(result, lastResult.current)) {
```

Comment:
It shouldn't be a problem, and avoiding a deep check is important for performance.

Because the data in the store is immutable, any change to some piece of data at a leaf requires all the parent references to change. For example, say the callback accesses (via a method call) `state.foo` from the store, then any change to an object under `state.foo` (eg. `state.foo.bar`) must also result in a change of the `state.foo` reference.